api documentation (#55)
chrisvdm/redwoodjs-stripe: README.md

`yarn rw dev`

> **Note**
>

> > The following steps won't be needed for much longer

Reviewer: Recommend you get rid of this as it's just describing how to use Stripe products instead of this plugin.

5. Checkout a subscription item using "user@test.com" as the email.

6. Configure Customer Portal in the [Stripe Dashboard](https://dashboard.stripe.com/subscriptions) 

7. Play around!

## Web-side API Reference

> After setting up your app via the setup command, you'll see that a demo page has been added to your app. This is intended as an in-app API reference—I'm assuming you would want your commerce project to look a different.

Reviewer: When writing these readme's try not to use first person. The project has several collaborators so it should reflect that.

Author: Good point!

### `<StripeProvider customer/>`

Uses React context to manage the StripeCart, store customer data and contain the plugin logic on the web side.

**Example 1**

```jsx
import { StripeProvider } from 'redwoodjs-stripe/web'

// ...

 <StripeProvider
        customer={{
          search: isLoggedIn ? `email: "${emailFromAuth}"` : '',
        }}
      >
        <Storefront/>
  </StripeProvider>
```

**Example 2**

> Routes.js

Reviewer: Assume that anyone seeing this is unfamiliar with Redwood concepts. Should link to some documentation about `routes.js` and why you'd pick this approach over the one in example 1.

```jsx
import { StripeProvider } from 'redwoodjs-stripe/web'

// ...

<Set wrap={StripeProvider} customer={{search: isLoggedIn ? `email: "${emailFromAuth}"` : ''}}>
  <Route path="/category" page={CategoryPage} name="category" />
  <Route path="/product" page={ProductPage} name="product" />
  // ...
</Set>
```

`customer` has the following shape:

> ```ts

Reviewer: This and all other types should ideally go into a `types.d.ts` file somewhere so they can be automatically imported and used in VSCode.

type Customer = {
  search: string | ''
}
```

> `search` accepts a [querystring](https://stripe.com/docs/search#search-query-language) that is used to search for a Stripe Customer. The idea is to use data of your authenticated user to build the querystring. In most cases your authentication provider will provide you with an email address that you will be able to use to map to a Stripe Customer.

Reviewer: It's not really a querystring since that's part of a URL. Maybe:
> `search` accepts a Search API query that is used...

> ### `useStripeCheckout()`

Reviewer: Let's not use the function name itself as a section title as it's hard to grok. Try something like this:

> ### Stripe Checkout hook
> Redirect to Stripe's hosted Checkout by using the `useStripeCheckout` hook.

    cancelUrl: 'http://localhost:8910/stripe-demo?success=false'
  })
 ```
 
`session` has the following shape:

```ts
type Session = {
  mode?: string       // optional if cart items have 'type' as part their objects, else defaults to 'once-off'
  cart?: Product[]    // 'cart' is not required when using built-in cart
  cancelUrl: string
  successUrl: string
}
```

### `useStripeCart()`

Hook to manage the Stripe Cart. Returns an object with the cart api (described below).

**Example**
```js
import { useStripeCart } from 'redwoodjs-stripe/web'
// ...
const { 
  cart, 
  clearCart, 
  addToCart, 
  removeFromCart, 
  editCartItem
} = useStripeCart()
```

### Cart API

#### `cart`

An array of Product objects. Each `product` has the following shape:

```ts
type Product = {
  id: string
  name?: string
  images?: string[]
  description?: string
  price?: number
  type?: string
}
```

Each object _must_ contain an `id`. This `id` corresponds to the StripePrice's `id` when using the generated StripeProductsCell.

If you're checking out with both recurring and once-off items, `type` needs to be included in the object.

> #### `clearCart()`

Reviewer: This and the following function calls would work better as a table I think.

Returns a `redirectToStripeCustomerPortal` function. 

```js
import { useStripeCustomerPortal } from 'redwoodjs-stripe/web'
// ...
const { redirectToStripeCustomerPortal } = useStripeCustomerPortal()
```

#### `redirectToStripeCustomerPortal(portalSession, skipAuth = false)`

Creates a Stripe Customer Portal Session and redirects to Stripe's Customer Portal page.

```js
await redirectToStripeCustomerPortal(
  {
    return_url: 'http://localhost:8910/stripe-demo',
  },
  true
)
```

`portalSession` has the following shape ([Stripe Customer Portal API reference](https://stripe.com/docs/api/customer_portal/sessions/create)):

```ts
type PortalSession = {
  configuration?: config
  locale?: string
  on_behalf_of?: string
  return_url: string
}
```

Set the second parameter, `skipAuth`, to `true` when testing the Customer Portal redirect if you have not yet set up user authentication in your app. Once authentication is set up please remove the argument.

#### `createStripeCustomerPortalConfig(portalConfig)`

Creates a [Customer Portal Configuration](https://stripe.com/docs/api/customer_portal/configuration) object.

```js

```ts
type PortalConfig = {

>   // TODO

Reviewer: Maybe just remove this till it's been added?

```js
import StripeProductsCell from 'src/components/StripeProductsCell/StripeProductsCell'

// ...

<StripeProductsCell
  params={{
    productParams: { active: true },
    priceParams: { type: 'recurring' },
  }}
/>
```

`priceParams` has the following shape (for more information, look at the [Stripe Price list API documentation](https://stripe.com/docs/api/prices/list))

```ts
type priceParams = {
  active?: boolean
  currency?: string
  product?: string
  created?: stripeCreated
  ending_before?: string
  limit?: number
  lookup_keys?: string
  recurring?: stripeRecurringPrice
  starting_after?: string
  type?: string
}
```

`productParams` has the following shape (for more information, look at the [Stripe Products list API documentation](https://stripe.com/docs/api/products/list))

```ts
type productParams = {
  active?: boolean
  created?: stripeCreated
  ending_before?: string
  ids?: string[]
  limit?: number
  shippable?: boolean
  starting_after?: string
  url?: string
}
```

The Cell has it's own UI for illustrative purpose. You should replace the UI with your own.

## Api-side API Reference

### Webhook handling

> As part of the setup process, a `/stripeWebhook` function is generated. It uses the plugin handler function to verify and manage webhook events.

Reviewer: Some might prefer their endpoints be snake case rather than camel case. Maybe throw in a line telling people how to edit that?

#### `handleStripeWebhooks(event, context, webhookObject[, verify])`

This function verifies and manages Stripe webhook events. `webhookObj` is an object of

```js
import { handleStripeWebhooks } from 'redwoodjs-stripe'
// ...

> const { results } = await handleStripeWebhooks(

Reviewer: This appears to be missing [webhook signature verification](https://stripe.com/docs/webhooks/signatures), which is a pretty big deal. If that's not planned for this RC can we add a comment addressing it? (and add it to the roadmap).

Author: It's just hidden inside `handleStripeWebhooks`. I'll document it a bit better. Thanks for this catch.

    event,
    context,
    {
      'customer.updated': async (e) => {
        updateUser(e)
      },
      'payment_intent.succeeded': (e) => {
        sendCustomerThankYouEmail(e)
      },
    }
  )
```

> #### Installing Stripe CLI

Reviewer: I'd remove this entirely and just link to the [Stripe CLI repo](https://github.com/stripe/stripe-cli), which has install instructions for every supported platform.

I'd just keep the note that you need to put your webhook signing secret in your `.env` file and show the webhook testing instructions.

Author: Gotcha!

```

### Schemas

Schemas are imported from the plugin. You can view the sdl files [here](https://github.com/chrisvdm/redwoodjs-stripe/tree/main/src/api/schemas). Although I tried to have the sdls match the [Stripe objects](https://stripe.com/docs/api), there may be times you need something more. For that reason my sdl files follow a naming convention so that you can add your own custom Stripe sdl files without there being a conflict. 

The following Stripe API objects have sdl files: 
- [Customers](https://stripe.com/docs/api/customers)
- [Subscriptions](https://stripe.com/docs/api/subscriptions)
- [Customer Portal](https://stripe.com/docs/api/customer_portal)
- [Payment Methods](https://stripe.com/docs/api/payment_methods)
- [Coupons](https://stripe.com/docs/api/coupons)
- [Discounts](https://stripe.com/docs/api/discounts)

## Current progress
 
You can take a look at the [Roadmap](https://github.com/chrisvdm/redwoodjs-stripe/issues/1) for this project for more details and a rough timeline. For a quick status update look below. 

**STATUS:** Documentation 📖 + bug squashing 🪳🔨

**ETA** August 2022 🤞

## How you can help

> - Try out this very unstable release and give [feedback](https://github.com/chrisvdm/redwoodjs-stripe/discussions/60)

Reviewer: Take out the "very unstable" part, it clearly works :)

`yarn rw dev`

_The following steps won't be needed for much longer_

> 5. Checkout a subscription item using "user@test.com" as email.

Reviewer 2: ```suggestion
5. Checkout a subscription item using "user@test.com" as the email.
```

6. Configure Customer Portal in the [Stripe Dashboard](https://dashboard.stripe.com/subscriptions) 

7. Play around!

## Web-side API Reference

After setting up your app via the setup command, you'll find a demo page has been added to your app. This is intended as an in-app API reference, I'm assuming you would want your commerce project to look a different.

### `<StripeProvider customer/>`

> Uses a provider to manage the StripeCart, store customer data and contain the plugin logic on the web side.

Reviewer 2: ```suggestion
Uses React context to manage the StripeCart, store customer data and contain the plugin logic on the web side.
```

This function verifies and manages Stripe webhook events. `webhookObj` is an object of

```js
import { handleStripeWebhooks } from 'redwoodjs-stripe'
// ...
const { results } = await handleStripeWebhooks(
    event,
    context,
    {
      'customer.updated': async (e) => {
        updateUser(e)
      },
      'payment_intent.succeeded': (e) => {
        sendCustomerThankYouEmail(e)
      },
    }
  )
```

#### Installing Stripe CLI

In order to take advantage of webhooks and Stripe webhook events, it is advisable that you install the [Stripe CLI](https://stripe.com/docs/cli) on your machine. You can follow the instructions below if you are using homebrew or follow [Stripe's installation guide](https://stripe.com/docs/stripe-cli#install).

`brew install stripe/stripe-cli/stripe`

Log into your Stripe account

`stripe login`

You will be given a webhook secret. Make sure it matches the key in your `.env` file under `STRIPE_WEBHOOK_KEY`

#### Test Stripe webhooks locally

1. Listen for webhook events locally

`stripe listen --forward-to localhost:8911/stripeWebhook`

2. Trigger webhook events locally in terminal

`stripe trigger customer.updated`

### Services

> Services used in the plugin can also be imported into your project for you to build custom services for those more complicated use-cases.

Reviewer 2: ```suggestion
Services used in the plugin can also be imported into your project to build custom Services for more-complicated use cases.
```

- Webhook handling 
- LoFi Cart

## Thank you

> The teams from [Stripe](https://stripe.com/) and [RedwoodJS](https://redwoodjs.com/) who have supported this project from day one.

Reviewer: Maybe include the contributors who aren't affiliated with either Stripe or Redwood?

Author: I'm looking into tools that automatically display contributors' mugshots
